src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+-- code that dynamically counts the number of times every index is
+-- actually accessed.
+--
+-- - An Access becomes a function consuming a buffer reference
+--   and returning unit.  The buffer has the same type as the
+--   object whose accesses we are analyzing, except for integer
+--   leaves.
+--
+-- - For every control path of the source program, there is a control
+--   path of the Access program that accumulates at least as many
+--   counts into each element of the buffer as the number of times the
+--   source program read that element.
+--
+-- - `Location exprs ct` turns into accumulation
+--     \ref. ref!<exprs> += ct
+--     Note that this += must be understood as pointwise if
+--     the source indexing expression is not fully saturated.
+--
+-- - `Any opts` turns into case
+--     \ref. index = randint $ length opts
+--       case index of
+--         1 -> compile opt1 ref
+--         2 -> compile opt2 ref
+--         ...
+--
+-- - `All parts` turns into sequentialization
+--     \ref.
+--       compile part1 ref
+--       compile part2 ref
+--       ...
+--
+-- - `ForEach b sub` is iteration
+--     \ref. for b in (type of b).
+--       compile sub ref
+
+-- === Use ===
+
+-- The Use structure is a "data space" representation of accessing
+-- arrays.
+--
+-- Note that a Use can still depend on iteration-space indices (from
+-- the `iter` scope)---it just doesn't have internal binders that are
+-- meant to be interpreted in iteration space.
+--
+-- A Use implicitly iterates over as many dimensions of the accessed
+-- object as there are nesting levels of At constructors in the Use.
+-- We do not use explicit binders for this because they always appear
+-- in the same nesting order and can always be pulled to the outermost
+-- layer, so it's simplest to leave them implicit.
+
+data Use (iter::S) =
+  -- The object is accessed in its entirety
+    Const Count
+  -- The object is accessed according to the third argument, but only
+  -- where the first argument equals the second.
+  | At (IxExpr iter) (Use iter)
+  | Max [Use iter]
+  | Sum [Use iter]
+  deriving (Eq, Show)
+
+-- The semantics of Use are similar to those for Access.  Here is the
+-- compilation that turns a Use into a program that dynamically counts
+-- how many times each index is used.
+--
+-- Note that now IxExpr permits the `IxAll` constructor, so an
+-- `IxExpr` represents a subset of indices rather than a single index.
+--
+-- - `Const ct` becomes accumulation
+--     \ref. ref += ct
+--     Also polymorphic in the ref type, applying elementwise
+--
+-- - `Max alts` becomes case
+--     \ref. case randint (length alts) of
+--       1 -> compile alt1 ref
+--       2 -> compile alt2 ref
+--       ...
+--
+-- - `Sum parts` becomes sequencing
+--     \ref.
+--       compile alt1 ref
+--       compile alt2 ref
+--       ...
+--
+-- - `At expr use` becomes testing and slicing
+--     \ref. for b in <next object dim>.
+--       if b in expr then (compile use ref!b)
```

Comment:
what does `b in expr` mean?

Reply:
The expression in general represents a set of positions.  Comment added to the `At` constructor of `Use`.

Reply:
Yes.  But in a case like `expr == Inject 0 IxAll`, this if is really testing "is `b` a `Left`?"